### DISCUSSION TITLE
How does this library detect any new device/models that arrive ?

### DISCUSSION THREAD

**Discussion by AdarshPanigrahi** [nielsbasjes/yauaa#904] [Q&A] [answered]
Which class is responsible for detecting new devices/models have arrived or not ?

**Comment by nielsbasjes** [comment]
None, The devices and models are extracted from the useragent.
If it is not in there it cannot be extracted.
Do you have an example of an incorrectly classified/extracted device?

**Reply by nielsbasjes** [reply]
To clarify a bit more: For a limited number of device brands (i.e. Apple and Amazon) I do have a lookup to map a device id into something more readable. For those I have scripting to automatically update it.

**Reply by AdarshPanigrahi** [reply]
Okay.I think you are referring to something like AppleTypes.yaml. I wanted to know how do you update this list if a new apple device comes up ?

**Reply by nielsbasjes** [reply]
I have a few scripts to fetch and compare with external sources which I periodically run: https://github.com/nielsbasjes/yauaa/tree/main/analyzer/src/main/code-gen/UserAgents/AppleTypes/

**Reply by AdarshPanigrahi** [reply]
So while performing detection for such an device, does it first lookup to the map and on failure tries to extract info from the user agent or vice-versa ?

**Reply by nielsbasjes** [reply]
Yes, and the lookup is done using information from the user agent.

**Reply by AdarshPanigrahi** [reply]
So if  want to use the library with my application which is on production, how do I ensure the newly added devices(e.g Apple devices) are updated also on my end and they can be handled if a request comes up ?

**Reply by nielsbasjes** [reply]
You should use a tool like renovate or dependabot to automatically update your system every time I release an updated version.

**Reply by AdarshPanigrahi** [reply]
Each time new useragents are added to the library, do you release an updated version of the library ?

**Reply by nielsbasjes** [reply]
Yes. And if you use something like [Renovate](https://renovatebot.com/) or [Dependabot](https://github.com/dependabot) you will get the latest versions soon after I have released them.

**Comment by nielsbasjes** [comment, marked as answer]
So simply put: Just automatically update your software to use the latest version and it is updated automatically for you.
